packages/devtools-proxy-support/src/proxy-agent.ts

```diff
@@ -0,0 +1,210 @@
+// (hence the license notice above), with core differences being that
+// this module uses a different `getProxyForUrl` signature for more flexibility
+// and loads the individual agents it defers to lazily.
+// Relevant pull requests have been linked in-line.
+
+import * as http from 'http';
+import * as https from 'https';
+import { URL } from 'url';
+import { LRUCache } from 'lru-cache';
+import type { AgentConnectOpts } from 'agent-base';
+import { Agent } from 'agent-base';
+import createDebug from 'debug';
+import type { PacProxyAgentOptions } from 'pac-proxy-agent';
+import type { PacProxyAgent } from 'pac-proxy-agent';
+import type { HttpProxyAgentOptions } from 'http-proxy-agent';
+import type { HttpProxyAgent } from 'http-proxy-agent';
+import type { HttpsProxyAgentOptions } from 'https-proxy-agent';
+import type { HttpsProxyAgent } from 'https-proxy-agent';
+import type { SocksProxyAgentOptions } from 'socks-proxy-agent';
+import type { SocksProxyAgent } from 'socks-proxy-agent';
+
+const debug = createDebug('proxy-agent');
+
+type ValidProtocol =
+  | typeof HttpProxyAgent.protocols[number]
+  | typeof HttpsProxyAgent.protocols[number]
+  | typeof SocksProxyAgent.protocols[number]
+  | typeof PacProxyAgent.protocols[number];
+
+type AgentConstructor = new (
+  proxy: string,
+  proxyAgentOptions?: ProxyAgentOptions
+) => Agent;
+
+type GetProxyForUrlCallback = (
+  url: string,
+  req: http.ClientRequest
+) => string | Promise<string>;
+
+/**
+ * Shorthands for built-in supported types.
+ */
+// https://github.com/TooTallNate/proxy-agents/pull/327
+const wellKnownAgents = {
+  http: async () => (await import('http-proxy-agent')).HttpProxyAgent,
+  https: async () => (await import('https-proxy-agent')).HttpsProxyAgent,
+  socks: async () => (await import('socks-proxy-agent')).SocksProxyAgent,
+  pac: async () => (await import('pac-proxy-agent')).PacProxyAgent,
+} as const;
+
+/**
+ * Supported proxy types.
+ */
+export const proxies: {
+  [P in ValidProtocol]: [
+    () => Promise<AgentConstructor>,
+    () => Promise<AgentConstructor>
+  ];
+} = {
+  http: [wellKnownAgents.http, wellKnownAgents.https],
+  https: [wellKnownAgents.http, wellKnownAgents.https],
+  socks: [wellKnownAgents.socks, wellKnownAgents.socks],
+  socks4: [wellKnownAgents.socks, wellKnownAgents.socks],
+  socks4a: [wellKnownAgents.socks, wellKnownAgents.socks],
+  socks5: [wellKnownAgents.socks, wellKnownAgents.socks],
+  socks5h: [wellKnownAgents.socks, wellKnownAgents.socks],
+  'pac+data': [wellKnownAgents.pac, wellKnownAgents.pac],
+  'pac+file': [wellKnownAgents.pac, wellKnownAgents.pac],
+  'pac+ftp': [wellKnownAgents.pac, wellKnownAgents.pac],
+  'pac+http': [wellKnownAgents.pac, wellKnownAgents.pac],
+  'pac+https': [wellKnownAgents.pac, wellKnownAgents.pac],
+};
+
+function isValidProtocol(v: string): v is ValidProtocol {
+  return Object.keys(proxies).includes(v);
```

Comment:
Why can't we just do `!!proxies[v]`? I know it's not very important in this case, but just to make sure I understand this part of the code.

Reply:
Oh, literally just kept this as it's a bit closer to the original code. No strong preferences tbh

Reply:
Ah then nothing, I thought there was some rewrite. Then let's keep it as the original.